### PR TITLE
[SCF-To-Calyx] Bug Fixes

### PR DESCRIPTION
**Pull request by calebmkim**
Edit: the changes from #5526 are showing up under the "Files Changed". I think I just have to  updates the branch to show the updates to main to make them go away. 

Two small changes in `SCFToCalyx` (they're unrelated but both pretty small so I thought both could go in a single PR) 

1)  When we see a  memref w/o dimensions (e.g., `memref<i32>`), we should lower it in Calyx to a 1 dimensional memory of size 1, since Calyx memories must have explicitly defined dimensions. Before it was erroring out when trying to export to Calyx, complaining that Calyx doesn't support "0-dimensional" memories. 

2) This is a bit nuanced, but I also changed the way we compile binOps (e.g., multiply, divide, etc.) to be more correct. 
Previously, binOps were compiled into groups as follows:
```
group perform_binOp {
  binOp.go = 1'd1; 
  // assign to binOp.left and binOp.right 
  reg.write_en = binOp.done; 
  reg.in = binOp.out; 
  perform_binOp[done] = reg.done; 
}
```
However, this is technically triggering `binOp.go` for the entire group, which means that it's being triggered even for the cycle that we spend writing into the register. The binOp being triggered for an extra cycle normally doesn't matter, but it does sometimes affect the `binOp.out` value when we try to read from it later on. Specifically, I was running into tricky bugs when the `binOp` was a signed division, but this fixed it: 
```
group perform_binOp {
  binOp.go = !binOp.done ? 1'd1; 
  // everything else is the same 
}
```
Just as an FYI, this way of compiling groups (with the ` binOp.go = !binOp.done ? 1'd1;`) is the standard way to perform a "binOp then write to register" in Calyx. For example, the Dahlia->Calyx compiler [expects this behavior](https://github.com/cucapra/calyx/blob/d95f53ef7a7694cb422e8baf8a412211eb1d3b81/tests/frontend/dahlia/dot-func.expect#L44)

### CODE REVIEW

**Comment by calebmkim** [comment]
@mortbopet @mikeurbach Sorry to ping, but for some reason I can't request a review for this PR. 

Also @andrewb1999 idk if you've already seen these problems, but these might be problems for `LoopScheduleToCalyx` as well.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 1]
```diff
@@ -284,12 +293,23 @@ class BuildOpGroups : public calyx::FuncOpPartialLoweringPattern {
...
-           "and address assignment values");
-    for (auto address : enumerate(addressValues))
-      rewriter.create<calyx::AssignOp>(loc, addrPorts[address.index()],
-                                       address.value());
+    if (addressValues.empty()) {
+      assert(
+          addrPorts.size() == 1 &&
+          "We expected a 1 dimensional memory of size 1 because there were no "
+          "address assignment values");
+      // Assign 1'd0 to the address port.
```
Comment (mikeurbach) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:301]: Is it legal to use a `0'd0` in Calyx? Other pipelines (e.g. Chisel -> FIRRTL -> CIRCT) deal with a similar edge case, where a 1 element vector has a 0 width wire to index into it. I guess it just matters that this is aligned with what Calyx expects, and if it expects a 1 bit address, then I guess that's fine.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp [hunk 2]
```diff
@@ -259,9 +261,16 @@ class BuildOpGroups : public calyx::FuncOpPartialLoweringPattern {
...
+    // Set pipelineOp to hi as long as its done signal is not high.
+    // This prevents the pipelineOP from executing when we write into the
+    // register.
+    // To get !(pipelineOp.done) we do 1 xor pipelineOp.done
+    hw::ConstantOp c1 = createConstant(loc, rewriter, getComponent(), 1, 1);
     rewriter.create<calyx::AssignOp>(
-        loc, opPipe.getGo(),
-        createConstant(loc, rewriter, getComponent(), 1, 1));
+        loc, opPipe.getGo(), c1,
+        builder.create<comb::XorOp>(group.getLoc(), opPipe.getDone(), c1,
```
Comment (mikeurbach) [inline lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:272]: CIRCT has a helper for creating a "not" using an XOR: https://github.com/llvm/circt/blob/72672984a7fa4092d2295a010527d647c2893ae3/lib/Dialect/Comb/CombOps.cpp#L48

**Review by mikeurbach — COMMENTED** [review]
I think the changes here make sense. I have the same unfortunate comment that will always come up until we finish unifying SCF and LoopSchedule to Calyx: does LoopScheduleToCalyx need these changes as well?

If you rebase or merge main into this branch the other changes should go away.

File: lib/Conversion/SCFToCalyx/SCFToCalyx.cpp (hunk 1 above)
Comment (calebmkim) [reply lib/Conversion/SCFToCalyx/SCFToCalyx.cpp:300]: Yeah, Calyx doesn't allow a 0 width wire, so we have to make it 1 bit.

**Comment by calebmkim** [comment]
> I think the changes here make sense. I have the same unfortunate comment that will always come up until we finish unifying SCF and LoopSchedule to Calyx: does LoopScheduleToCalyx need these changes as well?

So for fix 1 (the memref one), I just added the same fix to LoopScheduleToCalyx.
For fix 2 (the binary operators one), I think Andrew was actually thinking of changing the primitives used in LoopScheduleToCalyx to use pipelined primitives (i.e., primitives that accept new inputs every cycle) and we're having a Calyx meeting about it soon, so may hold off on the fix now (although perhaps it will be useful to have this fix for the time being that Andrew hasn't had his PR). 

> If you rebase or merge main into this branch the other changes should go away.

Sorry, I don't have write privileges, so it seems like I cannot merge main into this branch by just pressing the a button. 
I'm not that good at using git either, so there's probably a way to do this from the command line but I'm not sure how, sorry.

**Comment by calebmkim** [comment]
So to answer the question about `LoopScheduletoCalyx`
(tl;dr is I don't think it's necessary to do the change to binOps in `LoopScheduletoCalyx` in this PR) 

Andrew needs the actual Calyx multiply op that `LoopScheduleToCalyx` uses to be different than the the multiply op that `SCFToCalyx` uses. 

The main thing is that `LoopScheduleToCalyx` uses a multiplier that holds its output value until its go signal is written into. Andrew needs a multiplier that can accept a new input every cycle, and only holds its output for a single cycle (and if you don't use it in that cycle, you've lost it). 
In other words, `SCFToCalyx` doesn't use a Calyx pipelined multiplier (a bit confusing since it's named `std_mult_pipe` in Calyx idk why) whereas `LoopScheduleToCalyx` needs a pipelined multiplier in Calyx. 

So would it fine to merge this?

**Comment by calebmkim** [comment]
Ok I must have messed something up in git. Honestly at this point it seems easier just to redo the entire thing and copy and paste the changes I made before. It'll probably take like 5 minutes to redo everything since it wasn't a big PR.
